### PR TITLE
[DependencyInjection] [ProxyManager] Use lazy-loading ghost object proxies when possible

### LINKED ISSUE #35345: ProxyManagerBridge should possibly use Lazy Loading Ghost Object Proxies instead of Lazy Loading Value Holder Proxies
I have created new issue to make it less doctrine specific.

Here is demonstration of different lazy proxies (modified [proxy-manager examples]( https://github.com/Ocramius/ProxyManager/tree/master/examples))

```php

use ProxyManager\Factory\LazyLoadingGhostFactory;
use ProxyManager\Factory\LazyLoadingValueHolderFactory;

class Foo
{
    public function compare(self $foo)
    {
        var_dump($foo === $this);
    }
}


$proxy = (new LazyLoadingValueHolderFactory())->createProxy(Foo::class, function (&$wrappedObject) {
    $wrappedObject = new Foo();
    return true;
});

$proxy->compare($proxy); // bool(false)

$proxy = (new LazyLoadingGhostFactory())->createProxy(Foo::class, fn() => true);
$proxy->compare($proxy); // bool(true)
```

First approach is used by Symfony. As you can see, it breaks identity and is more complicated to use. I suggest to use second method, which doesn't break identities. Would solve https://github.com/symfony/symfony/issues/35216 and issues related to it.

cc @stof

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #35345
| License       | MIT
| Doc PR        | -

This PR makes proxy-manager-bridge use a [ghost object proxy](https://github.com/Ocramius/ProxyManager/blob/2.15.x/docs/lazy-loading-ghost-object.md) by default.

We keep using a [value holder proxy](https://github.com/Ocramius/ProxyManager/blob/2.15.x/docs/lazy-loading-value-holder.md):
- when the lazy definition has the "proxy" tag, since this is the semantics of the tag
- when the definition declares an abstract type as a "class" (can happen when using a factory)
- when the proxied class is internal

When a ghost object is initialized, two code paths are possible:
- if the service is created by constructor, then we call `__construct()` explicitly on the uninitialized object;
- if the service is created by a factory, then we copy the state of the object returned by the factory into the ghost object. To do so, we use VarExporter.

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/php/services_dedup_lazy_proxy.php [hunk 1]
```diff
@@ -42,6 +42,13 @@ protected function createProxy($class, \Closure $factory)
         return $factory();
     }
 
+    protected function constructInstance($class, $proxy, ...$arguments)
+    {
+        get_parent_class($proxy) === $class ? method_exists($proxy, '__construct') && $proxy->__construct(...$arguments) : $proxy = new $class(...$arguments);
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/php/services_dedup_lazy_proxy.php]: should probably check `is_callable` too, to account for private constructors.

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Tests/Fixtures/php/services_dedup_lazy_proxy.php]: Actually at this point having a private constructor would be a service definition error, no need to check for it on the proxy codepath while we don't for the code path that uses `new`. Better make them both fail in the usual loud way :)

File: src/Symfony/Bridge/Doctrine/ManagerRegistry.php [hunk 2]
```diff
@@ -50,26 +50,38 @@ protected function resetService($name): void
...
-                if (isset($this->fileMap[$name])) {
-                    $wrappedInstance = $this->load($this->fileMap[$name], false);
-                } else {
-                    $wrappedInstance = $this->{$this->methodMap[$name]}(false);
+            $initializer = function (GhostObjectInterface $manager, string $method, array $parameters, &$initializer, array $properties) use ($load) {
+                $instance = $load($manager);
+                $initializer = null;
+
+                if ($instance !== $manager) {
+                    throw new \LogicException(sprintf('A lazy initializer should return the ghost object proxy it was given as argument, but an instance of "%s" was returned.', \get_class($instance)));
```
Comment (stof) [inline src/Symfony/Bridge/Doctrine/ManagerRegistry.php]: should use `get_debug_type` instead

Comment (nicolas-grekas) [reply src/Symfony/Bridge/Doctrine/ManagerRegistry.php]: fixed

**Comment by nicolas-grekas** [comment]
For what it's worth, this PR works like a charm on a real project :)

**Comment by nicolas-grekas** [comment]
I updated the PR to keep using a value holder when the proxied class is internal.

**Comment by nicolas-grekas** [comment]
Closing in favor of #46752

**Pull request closed without merging**